controller/k8s/metadata_api.go

```diff
@@ -272,6 +273,82 @@ func (api *MetadataAPI) GetOwnerKindAndName(ctx context.Context, pod *corev1.Pod
 	return strings.ToLower(parent.Kind), parent.Name, nil
 }
 
+// The set of resources that we look up in the indexer are fairly niche. They all must be able to:
+// - be a parent of another resource, usually a Pod
+// - have a parent resource themselves
+// Currently, this is limited to Jobs (parented by CronJobs) and ReplicaSets (parented by
+// Deployments, StatefulSets, argo Rollouts, etc.). This list may change in the future, but
+// is sufficient for now.
+var indexedParents = map[string]indexedParent{
+	"Job":        {Job, batchv1.SchemeGroupVersion.WithResource("jobs")},
+	"ReplicaSet": {RS, appsv1.SchemeGroupVersion.WithResource("replicasets")},
+}
+
+type indexedParent struct {
+	resource APIResource
+	gvr      schema.GroupVersionResource
+}
+
+// GetRootOwnerKindAndName returns the resource's owner's type and metadata, using owner
+// references from the Kubernetes API. Parent refs are recursively traversed to find the
+// root parent resource, at least as far as the controller has permissions to query.
+func (api *MetadataAPI) GetRootOwnerKindAndName(ctx context.Context, tm *metav1.TypeMeta, om *metav1.ObjectMeta, retry bool) (*metav1.TypeMeta, *metav1.ObjectMeta) {
```

Comment:
The destination controller endpoint_watcher and workload_watcher still use the old GetOwnerKindAndName function for getting owner metadata.  Does it make sense to continue to maintain both of these are should GetRootOwnerKindAndName completely replace GetOwnerKindAndName and have the watchers use this instead?